src/terrain/terrain_helper.hpp

```diff
@@ -82,20 +83,19 @@ grow_grass_recursive(Terrain& ter, std::set<Tile*> all_grass) {
     for (Tile* tile : all_grass) {
         // is the tile and edge
         bool is_source = false;
-        for (int x = -1; x < 2; x++)
-            for (int y = -1; y < 2; y++) {
-                // for adjacent in same layer
-                Tile* adjacent_tile = Ninos_horrible_function(ter, tile, x, y);
-                if (adjacent_tile == nullptr){
-                    continue;
-                }
-                // test if that tile is over the edge
-                // (in some cases: not solid, and others: solid and not grass)
-                if (edge_detector(adjacent_tile)) {
-                    is_source = true;
-                    break;
-                }
+        for (auto it = ter.get_tile_adjacent_iterator(
+                 ter.pos(tile),
+                 DirectionFlags::HORIZONTAL1 | DirectionFlags::HORIZONTAL2
+             );
+             !it.end(); it++) {
```

Comment:
Move the iterator outside:
```suggestion
        auto it = ter.get_tile_adjacent_iterator(
                 ter.pos(tile),
                 DirectionFlags::HORIZONTAL1 | DirectionFlags::HORIZONTAL2
        );
        for (; !it.end(); it++) {
```